src/formula/variant_private.cpp

```diff
@@ -0,0 +1,282 @@
+/*
+   Copyright (C) 2017 by the Battle for Wesnoth Project http://www.wesnoth.org/
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+
+   See the COPYING file for more details.
+*/
+
+#include "formula/variant.hpp"
+#include "formula/variant_private.hpp"
+
+#include "formula/callable.hpp"
+
+namespace game_logic
+{
+
+variant variant_int::build_range_variant(int limit) const
+{
+	const int len = std::abs(limit - value_) + 1;
+
+	std::vector<variant> res;
+	res.reserve(len);
+
+	for(int i = value_; res.size() != res.capacity(); value_ < limit ? ++i : --i) {
+		res.emplace_back(i);
+	}
+
+	return variant(&res);
+}
+
+std::string variant_decimal::to_string_impl(const bool sign_value) const
+{
+	std::ostringstream ss;
+
+	int fractional =  value_ % 1000;
+	int integer    = (value_ - fractional) / 1000;
+
+	if(sign_value) {
+		// Make sure we get the sign on small negative values.
+		if(integer == 0 && value_ < 0) {
+			ss << '-';
+		}
+	}
+
+	ss << integer << ".";
+
+	fractional = std::abs(fractional);
+
+	if(fractional < 100) {
+		if(fractional < 10) {
+			ss << "00";
+		} else {
+			ss << 0;
+		}
+	}
+
+	ss << fractional;
+
+	return ss.str();
+}
+
+variant_callable::variant_callable(const formula_callable* callable)
+	: callable_(callable)
+	, mutable_callable_(nullptr) // FIXME
+{}
+
+std::string variant_callable::get_serialized_string() const
+{
+	// TODO: make serialize return a string.
```

Comment:
I think it would be better to make formula_callable::serialize take the stringstream as a parameter (specifically, probably as an `std::ostream`).